File: machine/cpp/audio/output_resampler.cpp
#include "audio/output_resampler.h"

#include "machine/common/numeric.h"
#include "spec/audio/apu.h"
#include "machine/devices/audio/output_ring.h"

namespace bmsx {

void AudioOutputResampler::reset() {
	m_outputRate = 0;
	m_phase = 0.0;
	m_hasCurrent = false;
	m_hasNext = false;
}

auto AudioOutputResampler::pull(
	ApuOutputRing& ring,
	i16* output,
	size_t frameCount,
	i32 outputSampleRate
) -> size_t {
	if (m_outputRate != outputSampleRate) {
		reset();
		m_outputRate = outputSampleRate;
	}
	if (m_hasCurrent && ring.queuedFrames() != 0u && ring.firstFrameSequence() != m_lastSourceSequence + 1) {
		m_phase -= static_cast<i64>(m_phase);
		m_hasCurrent = false;
		m_hasNext = false;
	}
	const f64 sourceStep = static_cast<f64>(APU_SAMPLE_RATE_HZ) / static_cast<f64>(outputSampleRate);
	size_t outputIndex = 0u;
	size_t producedFrames = 0u;
	while (producedFrames < frameCount) {
		if (!prime(ring)) {
			break;
		}
		while (m_phase >= 1.0) {
			if (ring.queuedFrames() == 0u) {
				return producedFrames;
			}
			m_phase -= 1.0;
			m_currentLeft = m_nextLeft;
			m_currentRight = m_nextRight;
			readNext(ring);
		}
		const f64 left = static_cast<f64>(m_currentLeft)
			+ static_cast<f64>(m_nextLeft - m_currentLeft) * m_phase;
		const f64 right = static_cast<f64>(m_currentRight)
			+ static_cast<f64>(m_nextRight - m_currentRight) * m_phase;
		output[outputIndex] = static_cast<i16>(roundI32(left));
		output[outputIndex + 1u] = static_cast<i16>(roundI32(right));
		outputIndex += 2u;
		producedFrames += 1u;
		m_phase += sourceStep;
	}
	return producedFrames;
}

bool AudioOutputResampler::prime(ApuOutputRing& ring) {
	if (!m_hasCurrent) {
		if (ring.queuedFrames() == 0u) {
			return false;
		}
		m_lastSourceSequence = ring.firstFrameSequence();
		const u32 packed = ring.readFramePacked();
		m_currentLeft = static_cast<i16>(packed & 0xffffu);
		m_currentRight = static_cast<i16>(packed >> 16u);
		m_hasCurrent = true;
	}
	if (!m_hasNext) {
		if (ring.queuedFrames() == 0u) {
			return false;
		}
		readNext(ring);
	}
	return true;
}

void AudioOutputResampler::readNext(ApuOutputRing& ring) {
	m_lastSourceSequence = ring.firstFrameSequence();
	const u32 packed = ring.readFramePacked();
	m_nextLeft = static_cast<i16>(packed & 0xffffu);
	m_nextRight = static_cast<i16>(packed >> 16u);
	m_hasNext = true;
}

} // namespace bmsx

File: machine/ts/audio/output_resampler.ts
import { APU_SAMPLE_RATE_HZ } from '../spec/audio/apu';
import type { ApuOutputRing } from '../machine/devices/audio/output_ring';

export class AudioOutputResampler {
	private outputRate = 0;
	private phase = 0;
	private hasCurrent = false;
	private hasNext = false;
	private lastSourceSequence = 0;
	private currentLeft = 0;
	private currentRight = 0;
	private nextLeft = 0;
	private nextRight = 0;

	public reset(): void {
		this.outputRate = 0;
		this.phase = 0;
		this.hasCurrent = false;
		this.hasNext = false;
	}

	public pull(
		ring: ApuOutputRing,
		output: Int16Array,
		frameCount: number,
		outputSampleRate: number,
	): number {
		if (this.outputRate !== outputSampleRate) {
			this.reset();
			this.outputRate = outputSampleRate;
		}
		if (this.hasCurrent && ring.queuedFrames() !== 0 && ring.firstFrameSequence() !== this.lastSourceSequence + 1) {
			this.phase %= 1;
			this.hasCurrent = false;
			this.hasNext = false;
		}
		const sourceStep = APU_SAMPLE_RATE_HZ / outputSampleRate;
		let outputIndex = 0;
		let producedFrames = 0;
		outputFrames:
		while (producedFrames < frameCount) {
			if (!this.prime(ring)) {
				break;
			}
			while (this.phase >= 1) {
				if (ring.queuedFrames() === 0) {
					break outputFrames;
				}
				this.phase -= 1;
				this.currentLeft = this.nextLeft;
				this.currentRight = this.nextRight;
				this.readNext(ring);
			}
			const left = this.currentLeft + (this.nextLeft - this.currentLeft) * this.phase;
			const right = this.currentRight + (this.nextRight - this.currentRight) * this.phase;
			output[outputIndex] = Math.round(left);
			output[outputIndex + 1] = Math.round(right);
			outputIndex += 2;
			producedFrames += 1;
			this.phase += sourceStep;
		}
		return producedFrames;
	}

	private prime(ring: ApuOutputRing): boolean {
		if (!this.hasCurrent) {
			if (ring.queuedFrames() === 0) {
				return false;
			}
			this.lastSourceSequence = ring.firstFrameSequence();
			const packed = ring.readFramePacked();
			this.currentLeft = (packed << 16) >> 16;
			this.currentRight = packed >> 16;
			this.hasCurrent = true;
		}
		if (!this.hasNext) {
			if (ring.queuedFrames() === 0) {
				return false;
			}
			this.readNext(ring);
		}
		return true;
	}

	private readNext(ring: ApuOutputRing): void {
		this.lastSourceSequence = ring.firstFrameSequence();
		const packed = ring.readFramePacked();
		this.nextLeft = (packed << 16) >> 16;
		this.nextRight = packed >> 16;
		this.hasNext = true;
	}
}
